### PR TITLE
[PAC][lld][AArch64][ELF] Support signed GOT with tiny code model

### LINKED ISSUE #87984: [PAC][ELF][lld] Support signed GOT

### PR DESCRIPTION
**Pull request by kovdan01**
Depends on #114525

Support `R_AARCH64_AUTH_GOT_ADR_PREL_LO21` and `R_AARCH64_AUTH_GOT_LD_PREL19`
GOT-generating relocations. A corresponding `RE_AARCH64_AUTH_GOT_PC` member
of `RelExpr` is added, which is an AUTH-specific variant of `R_GOT_PC`.

### CODE REVIEW

**Comment by kovdan01** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113816?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#113816** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113816?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#113815** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113815?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @kovdan01 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lld/test/ELF/aarch64-got-relocations-pauth.s
Comment (ilovepi) [inline lld/test/ELF/aarch64-got-relocations-pauth.s:105]: I assume these are intentionally not matched. In that case, is there a good reason to keep them in the test?

Comment (kovdan01) [reply lld/test/ELF/aarch64-got-relocations-pauth.s]: This is not output to be checked and matched but comments helping to understand the contents of hex dump. I've changed the prefix to `##` so it's clear that it's a comment and not a special line like RUN/CHECK/etc. See e841e190df73a5cbc6639cb40c467623f1b953ac

Comment (ilovepi) [reply lld/test/ELF/aarch64-got-relocations-pauth.s]: A thanks for the explanation. I think I made a similar comment in one of the other PRs before I saw this. Feel free to ignore that/mark it resolved.

**Review by ilovepi — COMMENTED** [review]
Again, LGTM, but lets get another maintainer to take a look before landing.

**Comment by ilovepi** [comment]
> Again, LGTM, but lets get another maintainer to take a look before landing.

well, assuming presubmit is working. I see a number of test failures, ATM.

**Comment by kovdan01** [comment]
> I see a number of test failures, ATM.

@ilovepi Yes, these are currently expected - as mentioned in the PR description, it depends on #114525 (codegen support for the feature). The dependency isn't included in the Graphite PR stack since this does not cause merge conflicts (different parts of llvm are altered) - I find it easier to keep PRs separate if it's possible and just mention unmerged dependencies in description (for me, it makes updating PRs more robust).

File: lld/test/ELF/aarch64-got-relocations-pauth.s
Comment (MaskRay) [inline lld/test/ELF/aarch64-got-relocations-pauth.s:83]: `none` is also an OS component and is misleading when used together with `linux`.

Just use `aarch64` (or `aarch64-linux` when it's important to mention Linux, usually it's not)

File: lld/test/ELF/aarch64-got-relocations-pauth.s
Comment (MaskRay) [inline lld/test/ELF/aarch64-got-relocations-pauth.s:145]: No need for so many instructions.

We want very few instructions that look like simplified real-world codegen. While it's important to test two symbols to add some nuance, we can simplify the code here. Perhaps we just need

```
adr x0, :got_auth:bar
ldr x1, :got_auth:zed
```

**Review by MaskRay — APPROVED** [review]
LGTM, but the description should mention why R_AARCH64_AUTH_GOT_PC is added and the existing RelExpr member that is similar to it.

File: lld/test/ELF/aarch64-got-relocations-pauth.s
Comment (kovdan01) [reply lld/test/ELF/aarch64-got-relocations-pauth.s:133]: Thanks for suggestion, removed excessive instructions in 3c80d75c601db4574bd0edd6b89dbd332cab5f2e

File: lld/test/ELF/aarch64-got-relocations-pauth.s
Comment (kovdan01) [reply lld/test/ELF/aarch64-got-relocations-pauth.s]: Thanks for suggestion, changed to `aarch64` in 3c80d75c601db4574bd0edd6b89dbd332cab5f2e. Also applied to #113815

File: lld/test/ELF/aarch64-got-relocations-pauth.s
Comment (MaskRay) [inline lld/test/ELF/aarch64-got-relocations-pauth.s:84]: I might just use tiny1 tiny2 instead of local-tiny external-tiny, since we may test both preemptible and non-preemptible symbols.    local/external isn't accurate linker term anyway (local may mean the local binding)

Comment (kovdan01) [reply lld/test/ELF/aarch64-got-relocations-pauth.s]: Thanks for suggestion, fixed in 7d949458a8c0ce85e52f558159298090fa3bf529, and also applied the same to #113815

**Comment by kovdan01** [comment]
### Merge activity

* **Dec 18, 1:40 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/113816).

**Pull request merged**
